keps/sig-cloud-provider/azure/20190125-out-of-tree-azure.md

```diff
@@ -50,81 +52,156 @@ status: provisional
+
+### Risks and Mitigation
 
-TODO for SIG-Azure
+#### API throttling
 
-### Risks and Mitigations
+Before CCM, kubelet supports getting Node information by cloud provider's instance metadata service. This includes:
 
-TODO for SIG-Azure
+- NodeName
+- ProviderID
+- NodeAddresses
+- InstanceType
+- AvailabilityZone
+
+But with CCM, this is not possible anymore because the above functionalities have been moved to cloud controller manager.
+
+Since API throttling is a main issue for large clusters, we have added caches for Azure resources in KCM. Those caches would also be added to CCM. But even with caches, there would still be API throttling issues on cluster provisioning stages and nodes initialization durations would be much longer than KCM because of throttling.
+
+This issue is being tracked on [kubernetes/cloud-provider#30](https://github.com/kubernetes/cloud-provider/issues/30). Its status would be updated later when it's discussed through sig cloud-provider.
+
+#### Azure credential provider
+
+Azure credential provider is also depending on cloud provider codes. Though Azure Managed Service Identity (MSI) is a way to avoid explicit setting of credentials, MSI is not available on all cases (e.g. MSI may not be authorized to specific ACR repository).
+
+This issue is being tracked on [kubernetes/cloud-provider#13](https://github.com/kubernetes/cloud-provider/issues/13). Its status would be updated later when it's discussed through sig cloud-provider.
 
 ## Design Details
 
 ### Test Plan
 
 Azure Cloud Controller provider is reporting conformance test results to TestGrid as per the [Reporting Conformance Test Results to Testgrid KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/0018-testgrid-conformance-e2e.md).
- See [report](https://testgrid.k8s.io/sig-azure-master#azure-master-conformance) for more details.
+
+See [report](https://testgrid.k8s.io/sig-azure-master#Summary) for more details.
 
 ### Graduation Criteria
 
-TODO for SIG-Azure
+- Azure cloud controller manager is moving to GA
+  - Feature compatible with KCM
+  - Conformance tests are passed and published to testgrid
+- CSI drivers for AzureDisk/AzureFile are moving to GA
+  - Feature compatible with KCM
+  - Conformance tests are passed and published to testgrid
+- Azure credential provider is still supported in Kubelet
+  - Feature compatible with KCM
+  - Conformance tests are passed and published to testgrid
 
 ### Upgrade / Downgrade Strategy
 
-TODO for SIG-Azure
+Upgrade/Downgrade Azure cloud controller manager and CSI drivers together with other master components. The versions for Azure cloud controller manager and CSI drivers should be chosen according to the version skew strategy below.
 
 ### Version Skew Strategy
 
-TODO for SIG-Azure
+For each Kubernetes minor releases (e.g. v1.15.x), dedicated Azure cloud controller manager and CSI drivers would be released.
```

Comment:
CSI driver should not depend on k8s release since CSI API spec is stable now, it has its own release schedule, what left to do is testing on each major release IMO.

Reply:
Right, not directly. it is actually depending on the CSI versions, and each Kubernetes release would have a list of supported CSI versions.

Reply:
fixed now